# Исследование надежности заемщиков


Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

**Описание данных**

* children — количество детей в семье
* days_employed — общий трудовой стаж в днях
* dob_years — возраст клиента в годах
* education — уровень образования клиента
* education_id — идентификатор уровня образования
* family_status — семейное положение
* family_status_id — идентификатор семейного положения
* gender — пол клиента
* income_type — тип занятости
* debt — имел ли задолженность по возврату кредитов
* total_income — ежемесячный доход
* purpose — цель получения кредита


## 1.Загрузка и изучение данных.

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/data/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/data/data.csv')
except:
    data = pd.read_csv('https://code3/data/data.csv)

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## 2.Предобработка данных.

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [20]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [21]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [22]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

## 3. Исследование данных.

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [24]:
#создадим функцию, которая распределит по категориям заемщиков в зависимости от колличества детей,
def family_category(children):
    if children < 1:
        return 'нет детей'
    if children == 1:
        return 'семья с 1 ребенком'    
    if children == 2:
        return 'семья с 2 детьми'
    if children > 2:
        return 'многодетная семья'
#добавим в таблицу новую колонку с этими данными    
data['family_category'] = data['children'].apply(family_category)
#проверим результат
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0     926.185831         27    высшее             0   
6         0    2879.202052         43    высшее             0   
7         0     152.779569         50   среднее             1   
8         2    6929.865299         35    высшее             0   
9         0    2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose total_income_category          purpose_category  \
0               покупка жилья                     B  операции с недвижимостью   
1     приобретение автомобиля                     C    операции с автомобилем   
2               покупка жилья                     C  операции с недвижимостью   
3  дополнительное образование                     B     получение образования   
4             сыграть свадьбу                     C        проведение свадьбы   
5               покупка жилья                     B  операции с недвижимостью   
6           операции с жильем                     B  операции с недвижимостью   
7                 образование                     C     получение образования   
8       на проведение свадьбы                     C        проведение свадьбы   
9     покупка жилья для семьи                     C  операции с недвижимостью   

      family_category  
0  семья с 1 ребенком  
1  семья с 1 ребенком  
2           нет детей  
3   многодетная семья  
4           нет детей  
5           нет детей  
6           нет детей  
7           нет детей  
8    семья с 2 детьми  
9           нет детей

In [25]:
#создадим сводную таблицу с новыми данными:
data_children = data.pivot_table(index='family_category', columns='debt', values='gender', aggfunc='count')
#создаем столбцы
data_children.columns = ['no_debt', 'debt']
#считаем долю должников
data_children['percentage_debtors_%'] = data_children['debt'] / (data_children['debt'] + data_children['no_debt'])
#сохраняем результат в %
data_children['percentage_debtors_%'] = data_children['percentage_debtors_%'] * 100
#сортируем таблицу по возрастанию
data_children.sort_values(by='percentage_debtors_%', ascending=True)

no_debt  debt  percentage_debtors_%
family_category                                        
нет детей             13028  1063              7.543822
многодетная семья       349    31              8.157895
семья с 1 ребенком     4364   444              9.234609
семья с 2 детьми       1858   194              9.454191

**Вывод:** зависимость между количеством детей и возвратом кредита в срок есть. 
1. Заемщики, у которых нет детей имеют меньше просрочек по задолжностям, чем те у кого есть дети, (относительно 9.45% по задолжностям, 7.54%  - это на 22% меньше) причем у заемщиков с 2мя детьми просрочек больше чем у заемщиков с 1 ребенком, т е риск просрочек растет с количеством детей. 
2. У многодетных детей процент просрочек немного ниже - но тут и выборка ощитимо меньше - не факт что отражает корректно зависимость, можно подробнее рассмотреть. Возможно также что у многодетных семей выше доход или влияет соцподдержка.Эта категория значительно меньше берет кредитов чем остальные, возможно стоит провести исследование и маркетологам поработать над привлечением этой аудитории

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [26]:
#создадим сводную таблицу с новыми данными:
data_family_status = data.pivot_table(index='family_status', columns='debt', values='gender', aggfunc='count')
#создаем столбцы
data_family_status.columns = ['no_debt', 'debt']
#считаем долю должников
data_family_status['percentage_debtors_%'] = data_family_status['debt'] / (data_family_status['debt'] + data_family_status['no_debt'])
#сохраняем результат в %
data_family_status['percentage_debtors_%'] = data_family_status['percentage_debtors_%'] * 100
#сортируем таблицу по возрастанию
data_family_status.sort_values(by='percentage_debtors_%', ascending=True)

no_debt  debt  percentage_debtors_%
family_status                                             
вдовец / вдова             888    63              6.624606
в разводе                 1105    84              7.064760
женат / замужем          11334   927              7.560558
гражданский брак          3749   385              9.313014
Не женат / не замужем     2523   273              9.763948

**Вывод:** зависимость между семейным положением и возвратом кредита в срок заключается в следующем:
1. Заемщики состоящие или ранее состоящие в официальных отношниях имеют меньший процент просрочек по кредиту.
2. Вдовцы/вдовы чаще других категорий заемщиков отдают долги вовремя (относительно максимального 9.76% по задолжностям, 6.62% - это на 32% меньше) - возможно это связано с получением наследства.
3. Самая рискованная категория - не состоящие в отношениях заемщики - 9.76%

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Ранее создали функцию "categorize_income()", которая делит заемщиков по категориям дохода:
- 0–30000 — 'E'
- 30001–50000 — 'D'
- 50001–200000 — 'C'
- 200001–1000000 — 'B'
- 1000001 и выше — 'A'

In [27]:
#создадим сводную таблицу с новыми данными:
data_income = data.pivot_table(index='total_income_category', columns='debt', values='gender', aggfunc='count')
#создаем столбцы
data_income.columns = ['no_debt', 'debt']
#считаем долю должников
data_income['percentage_debtors_%'] = data_income['debt'] / (data_income['debt'] + data_income['no_debt'])
#сохраняем результат в %
data_income['percentage_debtors_%'] = data_income['percentage_debtors_%'] * 100
#сортируем таблицу по возрастанию
data_income.sort_values(by='percentage_debtors_%', ascending=True)# Ваш код будет здесь. Вы можете создавать новые ячейки.

no_debt  debt  percentage_debtors_%
total_income_category                                     
D                          328    21              6.017192
B                         4660   354              7.060231
A                           23     2              8.000000
C                        14568  1353              8.498210
E                           20     2              9.090909

**Вывод:** зависимость между уровнем дохода и возвратом кредита в срок заключается в следующем:
1. Самый низкий процент просрочек по кредиту у заемщиков с доходом 30001–50000.
2. Категории с самым высоким доходои и с самым низким - не информативны, выборка слишком мала, их не будем учитывать.
3. Заемщики с доходом 50001–200000 имеют ощутимо больше задожностей чем заемщики с доходом 30001–50000, при этом заемщики с доходом 30001–50000 берут кредиты в разы меньше категорий В и С

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

Ранее создали функцию "categorize_purpose", которая разделила цели кредита по категориям:
- 'операции с автомобилем',
- 'операции с недвижимостью',
- 'проведение свадьбы',
- 'получение образования'

In [28]:
#создадим сводную таблицу с новыми данными:
data_purpose = data.pivot_table(index='purpose_category', columns='debt', values='gender', aggfunc='count')
#создаем столбцы
data_purpose.columns = ['no_debt', 'debt']
#считаем долю должников
data_purpose['percentage_debtors_%'] = data_purpose['debt'] / (data_purpose['debt'] + data_purpose['no_debt'])
#сохраняем результат в %
data_purpose['percentage_debtors_%'] = data_purpose['percentage_debtors_%'] * 100
#сортируем таблицу по возрастанию
data_purpose.sort_values(by='percentage_debtors_%', ascending=True)# Ваш код будет здесь. Вы можете создавать новые ячейки.# Ваш код будет здесь. Вы можете создавать новые ячейки.

no_debt  debt  percentage_debtors_%
purpose_category                                             
операции с недвижимостью     9971   780              7.255139
проведение свадьбы           2130   183              7.911803
получение образования        3619   369              9.252758
операции с автомобилем       3879   400              9.347978

**Вывод:** 
1. Заемщики, берущие кредит на недвижимость, имеют меньше просрочек чем все остальные заемщики. 
2. Самые рискованные категории - образование и операции с автомобилем, по ним больше просрочек на 22 % относительно просрочек кредитов на недвижимость, возможно это связано также с возрастом аудитории - можно дополнительно исследовать это предположение.
3.  Заемщиков категории операций с недвижимостью более 50% относительно всех остальных

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Причины могут быть следущие - неверный ввод данных, сбой при извлечении и сохранении, "человеческий" фактор, технические проблемы

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Не для всех количественных переменных, но в большинстве случаев медиана отражает наиболее адекватное значение переменных при различных аномалиях и сильно выраженных отклонениях

## 4. Общий вывод.

#### В данном исследовании на этапе предообработки были внесены следущие изменения: ####
 
 1. В колонке "total_income" (ежемесячный доход) были выявлены пропуски, которые заполнили медианным значением, а также  заменили вещественный тип данных на целочисленный.
 2. В колонке "days_employed" (трудовой стаж в днях) также выявлены пропуски и у двух типов (безработные и пенсионеры) выявлены аномально большие значения (более 14% аномальных значений). Пропуски заполнили медианными значениями по каждому типу занятости, аномалии не исправляли, т к данные из этого столбца в исследовании не используются.
 3. В колонке "children" (количество детей в семье) обнаружены и удалены строки  с аномальными значениями (-1,20)
 4. В колонке "education"(образование) обнаружены и устанены дубликаты, все строки приведены к нижнему регистру.
 5. Данные на основании "total_income" (ежемесячный доход) сформировали по категориям:
 
    * 0–30000 — 'E'
    * 30001–50000 — 'D'
    * 50001–200000 — 'C'
    * 200001–1000000 — 'B'
    * 1000001 и выше — 'A'

 
 6. Данные на основании "purpose" (цель получения кредита) сформировали по категориям:
 
    * 'операции с автомобилем'
    * 'операции с недвижимостью'
    * 'проведение свадьбы'
    * 'получение образования'


 7. Данные на основании "children" (количество детей в семье) сформировали по категориям:
 
    * 'нет детей'
    * 'семья с 1 ребенком'
    * 'семья с 2 детьми'
    * 'многодетная семья'

 
#### При исследовании данных были выявлены следущие зависимости: ####
 1. Заемщики, у которых нет детей имеют меньше просрочек по задолжностям, чем те у кого есть дети.
 2. Заемщики, состоящие или ранее состоящие в официальных отношениях имеют меньший процент просрочек по кредиту, и этот критерий наиболее показателен по влиянию на выплаты в срок.
 3. Самый низкий процент просрочек по кредиту у заемщиков с доходом 30001–50000
 4. Заемщики, берущие кредит на недвижимость, имеют меньше просрочек чем все остальные заемщики
 
 **Наиболее эффективный заемщик - это человек в разводе, без детей, с доходом 30-50 тысяч, берущий кредит на недвижимость.** 
 
 В целом % задолжностей по всем категориям не сильно отличается - от 6 до 9% от числа всех заемщиков, но самая большая разница в значениях выявлена в исследовании по категориям семейного положения (разница относительно минимального и максимального значений - 32%), значит этот фактор наиболее показателен